proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,456 @@
+
+```swift
+extension Percentage {
+
+  func modify(
+    inSeconds seconds: Int,
+    block: @escaping (inout Clamped<Int>) -> Void
+  ) { ... }
+  
+}
+
+
+let myPercentage = Percentage(percent: 50)
+
+myPercentage
+  .modify(inSeconds: 3) { percent in
+    percent.wrappedValue = 100
+    //    ^~~~~~~~~~~~ 
+    // Again, we have to access
+    // count through 'wrappedValue'.
+  }
+```
+
+In fact, establishing custom behavior on closure parameters is really powerful. For example, if such a feature were supported, it could be used in conjunction with [Function Builders](https://github.com/apple/swift-evolution/blob/master/proposals/0289-function-builders.md) to expose data managed by a 'component' type. For instance, in SwiftUI [`ForEach`](https://developer.apple.com/documentation/swiftui/foreach) could leverage this feature to expose the mutable state of its data source to its 'content' closure. This would enable users to more easily work with the data source itself inside the closure instead of accessing the original property, which is particularly painful when working with collections, as shown in this example:
+
+```swift
+struct MyView: View {
+
+  // A simple Shopping Item that includes
+  // a 'quantity' and a 'name' property.
+  @State 
+  private var shoppingItems: [Item]
+
+
+  var body: some View {
+    ForEach(0 ..< shoppingItems.count) { index in
+  
+      TextField("Enter the item's name...", $shoppingItems[index].name)
+      
+    }
+  }
+  
+}
+```
+
+we would – with an appropriate initializer – be able to simplify the above code, reducing boilerplate as a result:
+
+```swift
+struct MyView: View {
+
+  @State 
+  private var shoppingItems: [Item]
+
+
+  var body: some View {
+    ForEach($shoppingItems) { @Binding shoppingItem in
+    
+      TextField("Enter the item's name...", $shoppingItem.name)
+      
+    }
+  }
+  
+}
+```
+
+
+## Proposed solution
+
+We propose to extend the contexts were application of property-wrapper types is permitted. Namely, application of such types will be allowed on function and closure parameters:
+
+
+## Detailed design
+
+Property wrappers are essentially sugar wrapping a given property with compiler synthesized code. This proposal retains this principle employing the following rules for transformation.
+
+### Property Wrappers on Function Parameters
+
+Function parameters marked with compatible property-wrapper custom attributes must conform to the following rules:
```

Comment:
@filip-sakel I noticed that you changed some of the language in these rules to be more consistent (thank you!), but I think some details got lost along the way. I wrote "a **set** of property wrapper custom attributes" to make it clear that these rules are generalized for property wrapper composition, e.g.

```swift
func test(@Wrapper @Wrapper @Wrapper param: Int) { ... }
``` 

Similarly, in the first rule below, it's important to say that **each** property wrapper type must provide a suitable `init(wrappedValue:)`.

Reply:
Does this look better?

---

Function parameters marked with a set of compatible property-wrapper custom attributes must conform to the following rules:

1. For initalization through `wrappedValue`'s type, all property-wrapper types must provide a suitable `init(wrappedValue:)`.
2. Each `wrappedValue` getter shall be `nonmutating`.
3. Default values for such parameters must be expressed in terms of the innermost `wrappedValue` type.

Reply:
Yes, thank you!